// pinkyjs/pinky.js
'use strict'

var path = require('path')
var childProcess = require('child_process')
var phantomjs = require('phantomjs-prebuilt')
var binPath = phantomjs.path

var _showModeResults = {}
var _showModeServer

var _spawnCounter = 0
var _spawnIdleTimeout
var _spawnIdleTimeoutTime = 1000

function deepMerge (a, b) {
  Object.keys(b).forEach(function (key) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      a[key] = deepMerge(deepMerge({}, a[key]), b[key])
    } else {
      a[key] = b[key]
    }
  })
  return a
}

// var childArgs = [
//   '--web-security=false', // enable xhr/xss
// 
//   // enable tor
//   '--proxy-type=socks5',
//   '--proxy=127.0.0.1:9050',
// 
//   path.join(__dirname, 'pinky.phantom.js')
// ]

function parseChildArgs (opts) {
  var args = []

  args.push('--web-security=$bool'.replace('$bool',
    !(opts.cors || opts.xhr || opts.xss || !opts.webSecurity)
  ))

  if (opts.tor) { // helper method for default tor setup
    opts.proxy = Object.assign({
      type: opts.tor.type || 'socks5',
      address: opts.tor.address || '127.0.0.1',
      port: opts.tor.port || '9050'
    }, opts.proxy || {})
  }

  if (opts.proxy) {
    if (opts.proxy.type) {
      args.push('--proxy-type=$type'.replace('$type', opts.proxy.type))
    }
    if (opts.proxy.address || opts.proxy.port) {
      args.push('--proxy=$addr:$port'
        .replace('$addr', opts.proxy.address || 'localhost') // localhost
        .replace('$port', opts.proxy.port || '80') // default tor port
      )
    }
  }

  args.push(path.join(__dirname, 'pinky.phantom.js'))
  return args
}

function log (str) {
  if (!process.env.DEBUG) return
  console.log('[pinky] ' + str)
}

function createSpawn (opts) {
  var env = {}
  var _killTimeout
  var _exiting = false
  var _exitedSuccessfully = false

  var listeners = {}
  var _spawnOpts = Object.assign({}, opts)

  var _idleTimeout
  var _idleTimeoutTime = _spawnOpts.idleTimeout || _spawnOpts.idleTimeoutTime || 3000
  var _pages = 0

  var _cameras = {}

  function listen (evt, callback) {
    listeners[evt] = listeners[evt] || []
    listeners[evt].push(callback)
  }

  function handle (evt, data) {
    log('  handling evt [' + evt + ']')
    if (listeners[evt] && listeners[evt].length > 0) {
      var callback = listeners[evt].splice(0, 1)[0]
      log('  callbacking evt [' + evt + ']')
      callback(data)
      log('  finishing evt [' + evt + ']')
    } else {
      log('no handler for evt: ' + evt)
    }
  }

  opts = Object.assign({
    cors: true,
    tor: false,
  }, opts)

  var childArgs = parseChildArgs(opts)

  var packageJson = require('./package.json')
  env.version = (packageJson['version'] || packageJson['VERSION']) || '?.?.?'

  var spawn = childProcess.spawn(binPath, childArgs, { env: env })
  spawn.stdin.setEncoding('utf8')
  spawn.stdout.setEncoding('utf8')

  log('spawn created')

  // cleanup spawn on main process exit
  process.on('exit', function () {
    spawn.kill()
  })
  log('attached process.exit cleanup listener')

  _spawnCounter++
  clearTimeout(_spawnIdleTimeout)
  spawn.on('exit', function () {
    _exiting = true
    _exitedSuccessfully = true
    clearTimeout(_killTimeout)
    log('SPAWN EXITED')
    _spawnCounter--
    clearTimeout(_spawnIdleTimeout)
    _spawnIdleTimeout = setTimeout(function () {
      // TODO
      if (_spawnCounter === 0) {
        console.log('spawnCounter ZERO idle timeout')
        _showModeServer && _showModeServer.close()
      }
    }, _spawnIdleTimeoutTime)
  })

  spawn.stderr.on('data', function (data) {
    console.error('spawn stderr: ' + data)
  })

  // listen for incoming data
  var buffer = ''
  spawn.stdout.on('data', function (data) {
    buffer += data.toString()
    var split = buffer.split('\n')
    buffer = split[split.length - 1]

    split.slice(0, -1).forEach(function (line) {
      var trim = line.trim()
      if (trim.length > 120) {
        log('trim: ' + trim.slice(0, 120) + '...')
      } else {
        log('trim: ' + trim)
      }
      if (trim && trim.length > 0) {
        try {
          var json = JSON.parse(trim)

          if (json.throwError) {
            throw new Error(json.error || json.throwError)
          }

          // if (json.error && (typeof callback === 'function')) {
          //   callback(new Error(json.error))
          // }
          if (json.error) {
            throw new Error(json.error)
          }

          if (json.log) return console.log(' log: ' + json.log)
          if (json.debug) return // ignore
          if (json.pageError)  {
            return log(json.pageError)
          }

          switch (json.type) {
            case 'exit':
              log(' -- phantomjs exited -- ')
              break
            case 'showModeRenderBase64': // special case (perpetual)
              console.log('got showMode render')
              var _results = _showModeResults[json.pageIndex] || []
              _results.push(json.result)
              _showModeResults.__latestPageIndex = json.pageIndex
              _showModeResults[json.pageIndex] = _results
              break
            case 'camera':
              var fn = _cameras[json.pageIndex]
              fn && fn(json.result)
              break
            case 'create':
            case 'open':
            case 'close':
            case 'wait':
            case 'eval':
            case 'evaluate':
            case 'render':
            case 'renderBase64':
            case 'sendEvent':
            case 'click':
            case 'keyboard':
            case 'heartbeat':
              handle(json.type, json)
          }
        } catch (err) {
          log('--- in catch ---')
          // if (typeof callback === 'function') callback(err)
          throw err
        }
      }
    })
  })

  log('added spawn.stdout data listener')
  log('returning and executing spawn')

  function createPage (opts, callback) {
    if (typeof opts === 'function') {
      callback = opts
      opts = {}
    }

    // sets up a http server to to receive and send intermittent
    // images from the page

    clearTimeout(_idleTimeout)

    if (opts.camera && (typeof opts.camera !== 'function')) {
      throw new Error('Error! camera function callback is not of type "function"')
    }

    var _opts =  {}
    deepMerge(_opts, opts)
    Object.keys(_opts).forEach(function (key) {
      var val = opts[key]
      if (typeof val === 'function') { // transform functions to booleans
        _opts[key] = true
      }
    })

    emit({
      type: 'create',
      opts: _opts
    })

    listen('create', function (data) {
      var _pageIndex = data.pageIndex

      // bind out camera callback function to receive events from our phantomjs script
      if (opts.camera) {
        _cameras[_pageIndex] = opts.camera
      }

      clearTimeout(_idleTimeout)

      function open (url, callback) {
        emit({
          pageIndex: _pageIndex,
          type: 'open',
          url: url
        })

        listen('open', function (data) {
          _pages++
          clearTimeout(_idleTimeout)
          callback(data.error || data.err)
        })
      }

      function wait (querySelector, callback, timeoutTime) {
        emit({
          pageIndex: _pageIndex,
          type: 'wait',
          querySelector: querySelector,
          timeoutTime: timeoutTime // defaults to 25 seconds
        })

        listen('wait', function (data) {
          callback(data.error || data.err)
        })
      }

      function render (fileName, callback) {
        if (typeof fileName !== 'string') {
          throw new Error('fileName missing')
        }

        emit({
          pageIndex: _pageIndex,
          type: 'render',
          fileName: fileName
        })

        listen('render', function (data) {
          callback(data.error || data.err, data.result)
        })
      }

      function renderBase64 (format, callback) {
        if (typeof format === 'function') {
          callback = format
          format = 'PNG' // default to png
        }

        emit({
          pageIndex: _pageIndex,
          type: 'renderBase64',
          format: format
        })

        listen('renderBase64', function (data) {
          callback(data.error || data.err, data.result)
        })
      }

      function keyboard (querySelector, keys, callback) {
        if (typeof keys === 'function' && typeof callback === 'undefined') {
          callback = keys
          keys = querySelector
          querySelector = undefined
        }

        if (typeof keys === 'string') {
          keys = [keys]
        }

        var count = 0
        var _callback = function (data) {
          count++
          if (count === keys.length) {
            callback(data.error || data.err, data.result)
          }
        }

        keys.forEach(function (key, index) {
          emit({
            pageIndex: _pageIndex,
            type: 'keyboard',
            querySelector: (index === 0) ? querySelector : undefined,
            key: key
          })

          listen('keyboard', function (data) {
            _callback(data)
          })
        })
      }

      function click (querySelector, opts, callback) {
        if (typeof opts === 'function') {
          callback = opts
          opts = {}
        }

        if (!opts.button) {
          opts.button = 'left'
        }

        if (!opts.event) {
          opts.event = 'click'
        }

        emit({
          type: 'click',
          pageIndex: _pageIndex,
          querySelector: querySelector,
          opts: opts
        })

        listen('click', function (data) {
          callback(data.error || data.err, data.result)
        })
      }

      function sendEvent (args, callback) {
        emit({
          pageIndex: _pageIndex,
          type: 'sendEvent',
          args: args
        })

        listen('sendEvent', function (data) {
          callback(data.error || data.err, data.result)
        })
      }

      function evaluate (evalCallback, argument, callback) {
        if (typeof argument === 'function') {
          callback = argument
          argument = undefined
        }

        emit({
          pageIndex: _pageIndex,
          type: 'evaluate',
          callback: evalCallback.toString(),
          argument: argument
        })

        listen('evaluate', function (data) {
          callback(data.error || data.err, data.result)
        })
      }

      function waitEvaluate (evalCallback, argument, callback, timeoutTime) {
        if (typeof argument === 'function') {
          timeoutTime = callback
          callback = argument
          argument = undefined
        }

        emit({
          pageIndex: _pageIndex,
          type: 'evaluate',
          callback: evalCallback.toString(),
          argument: argument,
          timeoutTime: timeoutTime // defaults to 25 seconds
        })

        listen('evaluate', function (data) {
          callback(data.error || data.err, data.result)
        })
      }

      function close (callback) {
        emit({
          pageIndex: _pageIndex,
          type: 'close'
        })

        listen('close', function (data) {
          console.log('  page close event listen callback')

          _pages--
          if (_pages === 0) {
            clearTimeout(_idleTimeout)
            _idleTimeout = setTimeout(function () {
              console.log('  idleTimeout triggered')
              exit()
            }, _idleTimeoutTime)
          }

          if (typeof callback === 'function') {
            callback(data.error || data.err, data.result)
          }
        })
      }

      callback(undefined, {
        open: open,
        close: close,
        wait: wait,
        render: render,
        renderBase64: renderBase64,
        sendEvent: sendEvent,
        click: click,
        keyboard: keyboard,
        eval: evaluate,
        evaluate: evaluate,
        waitEval: waitEvaluate,
        waitEvaluate: waitEvaluate,
        event: {
          key: {
            '0': 48,
            '1': 49,
            '2': 50,
            '3': 51,
            '4': 52,
            '5': 53,
            '6': 54,
            '7': 55,
            '8': 56,
            '9': 57,
            'A': 65,
            'AE': 198,
            'Aacute': 193,
            'Acircumflex': 194,
            'AddFavorite': 16777408,
            'Adiaeresis': 196,
            'Agrave': 192,
            'Alt': 16777251,
            'AltGr': 16781571,
            'Ampersand': 38,
            'Any': 32,
            'Apostrophe': 39,
            'ApplicationLeft': 16777415,
            'ApplicationRight': 16777416,
            'Aring': 197,
            'AsciiCircum': 94,
            'AsciiTilde': 126,
            'Asterisk': 42,
            'At': 64,
            'Atilde': 195,
            'AudioCycleTrack': 16777478,
            'AudioForward': 16777474,
            'AudioRandomPlay': 16777476,
            'AudioRepeat': 16777475,
            'AudioRewind': 16777413,
            'Away': 16777464,
            'B': 66,
            'Back': 16777313,
            'BackForward': 16777414,
            'Backslash': 92,
            'Backspace': 16777219,
            'Backtab': 16777218,
            'Bar': 124,
            'BassBoost': 16777331,
            'BassDown': 16777333,
            'BassUp': 16777332,
            'Battery': 16777470,
            'Bluetooth': 16777471,
            'Book': 16777417,
            'BraceLeft': 123,
            'BraceRight': 125,
            'BracketLeft': 91,
            'BracketRight': 93,
            'BrightnessAdjust': 16777410,
            'C': 67,
            'CD': 16777418,
            'Calculator': 16777419,
            'Calendar': 16777444,
            'Call': 17825796,
            'Camera': 17825824,
            'CameraFocus': 17825825,
            'Cancel': 16908289,
            'CapsLock': 16777252,
            'Ccedilla': 199,
            'Clear': 16777227,
            'ClearGrab': 16777421,
            'Close': 16777422,
            'Codeinput': 16781623,
            'Colon': 58,
            'Comma': 44,
            'Community': 16777412,
            'Context1': 17825792,
            'Context2': 17825793,
            'Context3': 17825794,
            'Context4': 17825795,
            'ContrastAdjust': 16777485,
            'Control': 16777249,
            'Copy': 16777423,
            'Cut': 16777424,
            'D': 68,
            'DOS': 16777426,
            'Dead_Abovedot': 16781910,
            'Dead_Abovering': 16781912,
            'Dead_Acute': 16781905,
            'Dead_Belowdot': 16781920,
            'Dead_Breve': 16781909,
            'Dead_Caron': 16781914,
            'Dead_Cedilla': 16781915,
            'Dead_Circumflex': 16781906,
            'Dead_Diaeresis': 16781911,
            'Dead_Doubleacute': 16781913,
            'Dead_Grave': 16781904,
            'Dead_Hook': 16781921,
            'Dead_Horn': 16781922,
            'Dead_Iota': 16781917,
            'Dead_Macron': 16781908,
            'Dead_Ogonek': 16781916,
            'Dead_Semivoiced_Sound': 16781919,
            'Dead_Tilde': 16781907,
            'Dead_Voiced_Sound': 16781918,
            'Delete': 16777223,
            'Direction_L': 16777305,
            'Direction_R': 16777312,
            'Display': 16777425,
            'Documents': 16777427,
            'Dollar': 36,
            'Down': 16777237,
            'E': 69,
            'ETH': 208,
            'Eacute': 201,
            'Ecircumflex': 202,
            'Ediaeresis': 203,
            'Egrave': 200,
            'Eisu_Shift': 16781615,
            'Eisu_toggle': 16781616,
            'Eject': 16777401,
            'End': 16777233,
            'Enter': 16777221,
            'Equal': 61,
            'Escape': 16777216,
            'Excel': 16777428,
            'Exclam': 33,
            'Execute': 16908291,
            'Explorer': 16777429,
            'F': 70,
            'F1': 16777264,
            'F10': 16777273,
            'F11': 16777274,
            'F12': 16777275,
            'F13': 16777276,
            'F14': 16777277,
            'F15': 16777278,
            'F16': 16777279,
            'F17': 16777280,
            'F18': 16777281,
            'F19': 16777282,
            'F2': 16777265,
            'F20': 16777283,
            'F21': 16777284,
            'F22': 16777285,
            'F23': 16777286,
            'F24': 16777287,
            'F25': 16777288,
            'F26': 16777289,
            'F27': 16777290,
            'F28': 16777291,
            'F29': 16777292,
            'F3': 16777266,
            'F30': 16777293,
            'F31': 16777294,
            'F32': 16777295,
            'F33': 16777296,
            'F34': 16777297,
            'F35': 16777298,
            'F4': 16777267,
            'F5': 16777268,
            'F6': 16777269,
            'F7': 16777270,
            'F8': 16777271,
            'F9': 16777272,
            'Favorites': 16777361,
            'Finance': 16777411,
            'Flip': 17825798,
            'Forward': 16777314,
            'G': 71,
            'Game': 16777430,
            'Go': 16777431,
            'Greater': 62,
            'H': 72,
            'Hangul': 16781617,
            'Hangul_Banja': 16781625,
            'Hangul_End': 16781619,
            'Hangul_Hanja': 16781620,
            'Hangul_Jamo': 16781621,
            'Hangul_Jeonja': 16781624,
            'Hangul_PostHanja': 16781627,
            'Hangul_PreHanja': 16781626,
            'Hangul_Romaja': 16781622,
            'Hangul_Special': 16781631,
            'Hangul_Start': 16781618,
            'Hangup': 17825797,
            'Hankaku': 16781609,
            'Help': 16777304,
            'Henkan': 16781603,
            'Hibernate': 16777480,
            'Hiragana': 16781605,
            'Hiragana_Katakana': 16781607,
            'History': 16777407,
            'Home': 16777232,
            'HomePage': 16777360,
            'HotLinks': 16777409,
            'Hyper_L': 16777302,
            'Hyper_R': 16777303,
            'I': 73,
            'Iacute': 205,
            'Icircumflex': 206,
            'Idiaeresis': 207,
            'Igrave': 204,
            'Insert': 16777222,
            'J': 74,
            'K': 75,
            'Kana_Lock': 16781613,
            'Kana_Shift': 16781614,
            'Kanji': 16781601,
            'Katakana': 16781606,
            'KeyboardBrightnessDown': 16777398,
            'KeyboardBrightnessUp': 16777397,
            'KeyboardLightOnOff': 16777396,
            'L': 76,
            'LastNumberRedial': 17825801,
            'Launch0': 16777378,
            'Launch1': 16777379,
            'Launch2': 16777380,
            'Launch3': 16777381,
            'Launch4': 16777382,
            'Launch5': 16777383,
            'Launch6': 16777384,
            'Launch7': 16777385,
            'Launch8': 16777386,
            'Launch9': 16777387,
            'LaunchA': 16777388,
            'LaunchB': 16777389,
            'LaunchC': 16777390,
            'LaunchD': 16777391,
            'LaunchE': 16777392,
            'LaunchF': 16777393,
            'LaunchG': 16777486,
            'LaunchH': 16777487,
            'LaunchMail': 16777376,
            'LaunchMedia': 16777377,
            'Left': 16777234,
            'Less': 60,
            'LightBulb': 16777405,
            'LogOff': 16777433,
            'M': 77,
            'MailForward': 16777467,
            'Market': 16777434,
            'Massyo': 16781612,
            'MediaLast': 16842751,
            'MediaNext': 16777347,
            'MediaPause': 16777349,
            'MediaPlay': 16777344,
            'MediaPrevious': 16777346,
            'MediaRecord': 16777348,
            'MediaStop': 16777345,
            'MediaTogglePlayPause': 16777350,
            'Meeting': 16777435,
            'Memo': 16777404,
            'Menu': 16777301,
            'MenuKB': 16777436,
            'MenuPB': 16777437,
            'Messenger': 16777465,
            'Meta': 16777250,
            'Minus': 45,
            'Mode_switch': 16781694,
            'MonBrightnessDown': 16777395,
            'MonBrightnessUp': 16777394,
            'Muhenkan': 16781602,
            'Multi_key': 16781600,
            'MultipleCandidate': 16781629,
            'Music': 16777469,
            'MySites': 16777438,
            'N': 78,
            'News': 16777439,
            'No': 16842754,
            'Ntilde': 209,
            'NumLock': 16777253,
            'NumberSign': 35,
            'O': 79,
            'Oacute': 211,
            'Ocircumflex': 212,
            'Odiaeresis': 214,
            'OfficeHome': 16777440,
            'Ograve': 210,
            'Ooblique': 216,
            'OpenUrl': 16777364,
            'Option': 16777441,
            'Otilde': 213,
            'P': 80,
            'PageDown': 16777239,
            'PageUp': 16777238,
            'ParenLeft': 40,
            'ParenRight': 41,
            'Paste': 16777442,
            'Pause': 16777224,
            'Percent': 37,
            'Period': 46,
            'Phone': 16777443,
            'Pictures': 16777468,
            'Play': 16908293,
            'Plus': 43,
            'PowerDown': 16777483,
            'PowerOff': 16777399,
            'PreviousCandidate': 16781630,
            'Print': 16777225,
            'Printer': 16908290,
            'Q': 81,
            'Question': 63,
            'QuoteDbl': 34,
            'QuoteLeft': 96,
            'R': 82,
            'Refresh': 16777316,
            'Reload': 16777446,
            'Reply': 16777445,
            'Return': 16777220,
            'Right': 16777236,
            'Romaji': 16781604,
            'RotateWindows': 16777447,
            'RotationKB': 16777449,
            'RotationPB': 16777448,
            'S': 83,
            'Save': 16777450,
            'ScreenSaver': 16777402,
            'ScrollLock': 16777254,
            'Search': 16777362,
            'Select': 16842752,
            'Semicolon': 59,
            'Send': 16777451,
            'Shift': 16777248,
            'Shop': 16777406,
            'SingleCandidate': 16781628,
            'Slash': 47,
            'Sleep': 16908292,
            'Space': 32,
            'Spell': 16777452,
            'SplitScreen': 16777453,
            'Standby': 16777363,
            'Stop': 16777315,
            'Subtitle': 16777477,
            'Super_L': 16777299,
            'Super_R': 16777300,
            'Support': 16777454,
            'Suspend': 16777484,
            'SysReq': 16777226,
            'T': 84,
            'THORN': 222,
            'Tab': 16777217,
            'TaskPane': 16777455,
            'Terminal': 16777456,
            'Time': 16777479,
            'ToDoList': 16777420,
            'ToggleCallHangup': 17825799,
            'Tools': 16777457,
            'TopMenu': 16777482,
            'Touroku': 16781611,
            'Travel': 16777458,
            'TrebleDown': 16777335,
            'TrebleUp': 16777334,
            'U': 85,
            'UWB': 16777473,
            'Uacute': 218,
            'Ucircumflex': 219,
            'Udiaeresis': 220,
            'Ugrave': 217,
            'Underscore': 95,
            'Up': 16777235,
            'V': 86,
            'Video': 16777459,
            'View': 16777481,
            'VoiceDial': 17825800,
            'VolumeDown': 16777328,
            'VolumeMute': 16777329,
            'VolumeUp': 16777330,
            'W': 87,
            'WLAN': 16777472,
            'WWW': 16777403,
            'WakeUp': 16777400,
            'WebCam': 16777466,
            'Word': 16777460,
            'X': 88,
            'Xfer': 16777461,
            'Y': 89,
            'Yacute': 221,
            'Yes': 16842753,
            'Z': 90,
            'Zenkaku': 16781608,
            'Zenkaku_Hankaku': 16781610,
            'Zoom': 16908294,
            'ZoomIn': 16777462,
            'ZoomOut': 16777463,
            'acute': 180,
            'brokenbar': 166,
            'cedilla': 184,
            'cent': 162,
            'copyright': 169,
            'currency': 164,
            'degree': 176,
            'diaeresis': 168,
            'division': 247,
            'exclamdown': 161,
            'guillemotleft': 171,
            'guillemotright': 187,
            'hyphen': 173,
            'iTouch': 16777432,
            'macron': 175,
            'masculine': 186,
            'mu': 181,
            'multiply': 215,
            'nobreakspace': 160,
            'notsign': 172,
            'onehalf': 189,
            'onequarter': 188,
            'onesuperior': 185,
            'ordfeminine': 170,
            'paragraph': 182,
            'periodcentered': 183,
            'plusminus': 177,
            'questiondown': 191,
            'registered': 174,
            'section': 167,
            'ssharp': 223,
            'sterling': 163,
            'threequarters': 190,
            'threesuperior': 179,
            'twosuperior': 178,
            'unknown': 33554431,
            'ydiaeresis': 255,
            'yen': 165
          }
        }
      })
    })
  } // eof createPage function

  function emit (json) {
    if (!_exiting) {
      // log(' -- emitting: ' + json.type)
      var line = JSON.stringify(json)
      spawn.stdin.write(line + '\n')
    } else {
      // console.log('exiting...')
    }
  }

  function heartbeat () {
    emit({
      type: 'heartbeat'
    })

    listen('heartbeat', function () {
      // schedule next heartbeat
      setTimeout(heartbeat, 33)
    })
  }
  setTimeout(function () {
    heartbeat()
  }, 100)

  function exit () {
    if (!_exiting) {
      emit({
        type: 'exit'
      })
      _exiting = true
      setTimeout(function () {
        if (!_exitedSuccessfully) {
          clearTimeout(_killTimeout)
          _killTimeout = setTimeout(function () {
            log('KILL TIMEOUT')
            spawn && spawn.kill()
          }, 3000)
        }
      }, 1000)
    }
  }

  return {
    _spawn: spawn, // underlying spawn process
    createPage: createPage,
    exit: exit
  }
}

function init (opts, callback) {
  if (typeof opts === 'function') {
    callback = opts
    opts = {}
  }
  var s = createSpawn()
  s.createPage(opts, function (err, page) {
    callback(err, page)
  })
  return s
}

init.createSpawn = createSpawn
init.startServer = startServer

function startServer (opts) {
  var http = require('http')

  opts = Object.assign({}, opts)

  var _format = 'jpeg'
  var _port = opts.port || 7000
  var _jpegBase64 = ''
  var _running = true

  var _server = http.createServer(function (req, res) {
    if (!_running) {
      req.connection.unref()
    } else {
      if (req.url.indexOf('latest') > 0) {
        res.write(_jpegBase64)
        res.end()
      } else {
        res.write(`
          <body style="margin: 0">
            <img id='screen' src="data:image/${_format};base64,blabla">
            <script>
            function update () {
              var req = new XMLHttpRequest()
              req.open('GET', '/latest', true)
              req.onload = function () {
                var data = req.responseText
                var img = document.getElementById('screen')
                img.src = "data:image/jpeg;base64,$data".replace('$data', data)
                setTimeout(update, 33)
              }
              req.onerror = function () {
                setTimeout(update, 500)
              }
              req.send()
            }

            update()
            </script>
          </body>
        `)
        res.end()
      }
    }
  }).listen(_port)
  console.log('pinkyjs server listening at localhost:' + _port)

  var _sockets = []
  _server.on('connection', function (socket) {
    _sockets.push(socket)
    socket.on('close', function () {
      var i =_sockets.indexOf(socket)
      _sockets.splice(i, 1)
    })
  })

  _showModeServer = {
    set: function (jpegBase64, meta) {
      _jpegBase64 = jpegBase64
    },
    close: function () {
      _running = false
      _server.close() // stop listening for new connections
      _sockets.forEach(function (socket) {
        socket.destroy() // destroy all remaining connections (and shut down http server)
      })
    }
  }

  return _showModeServer
}

module.exports = init
